### ISSUE TITLE
[BUG] [Chapter 15] Wrong dimension for returned sequence in MyRNN class

### ISSUE THREAD

**Issue by eisthf** [ageron/handson-ml2#457]
```python
class MyRNN(keras.layers.Layer):
    def __init__(self, cell, return_sequences=False, **kwargs):
        super().__init__(**kwargs)
        self.cell = cell
        self.return_sequences = return_sequences
        self.get_initial_state = getattr(
            self.cell, "get_initial_state", self.fallback_initial_state)
    def fallback_initial_state(self, inputs):
        return [tf.zeros([self.cell.state_size], dtype=inputs.dtype)]
    @tf.function
    def call(self, inputs):
        states = self.get_initial_state(inputs)
        n_steps = tf.shape(inputs)[1]
        if self.return_sequences:
            sequences = tf.TensorArray(inputs.dtype, size=n_steps)
        outputs = tf.zeros(shape=[n_steps, self.cell.output_size], dtype=inputs.dtype)
        for step in tf.range(n_steps):
            outputs, states = self.cell(inputs[:, step], states) # the shape of outputs is (batch, state_size)
            if self.return_sequences:
                sequences = sequences.write(step, outputs) # At this point, sequences is a python list containing n_step steps of (batch, state_size)
        if self.return_sequences:
            return sequences.stack() # Stacking is done along the axis=0, therefore the resulting shape should be (n_step, batch, state_size)
        else:
            return outputs
```

Please refer to the above comments in the respective lines. `self.cell(inputs[:, step], states)` produces the output of shape `(batch, state_size)` . `sequences = sequences.write(step, outputs)` keeps adding the outputs `n_step` times. After the `for` loop, `sequences` contains `[(batch, state_size), (batch, state_size), ... , (batch, state_size)]` with a total of `n_step` items in it.
`sequences.stack()` stacks the items along the axis=0, so the resulting tensor shape should be `(n_step, batch, state_size)`. However since batch first convention is being followed in tensorflow RNN cell such as  `tf.SimpleRNNCell`, the code should be modified to `return tf.transpose(sequence.stack(), (1,0,2))`  

The following code also supports my argument; 
https://www.tensorflow.org/tutorials/structured_data/time_series

```python
def call(self, inputs, training=None):
  # Use a TensorArray to capture dynamically unrolled outputs.
  predictions = []
  # Initialize the LSTM state.
  prediction, state = self.warmup(inputs)

  # Insert the first prediction.
  predictions.append(prediction)

  # Run the rest of the prediction steps.
  for n in range(1, self.out_steps):
    # Use the last prediction as input.
    x = prediction
    # Execute one lstm step.
    x, state = self.lstm_cell(x, states=state,
                              training=training)
    # Convert the lstm output to a prediction.
    prediction = self.dense(x)
    # Add the prediction to the output.
    predictions.append(prediction)

  # predictions.shape => (time, batch, features)
  predictions = tf.stack(predictions)
  # predictions.shape => (batch, time, features)
  predictions = tf.transpose(predictions, [1, 0, 2])
  return predictions
```

Because `tf.transpose` returns a (transposed) copy of the original data, it takes more time than without it. I wonder if there is a way to create a correct dimension without copying in this situation.

**Comment by ageron** [comment]
Hi @eisthf ,

Thanks a lot for your feedback. Yikes, the `MyRNN` class actually contained no less than 4 errors (!!!), I'm surprised it produced no exception! My sincere apologies for these errors, I should have checked more thoroughly. I guess I just ran the model, saw that it didn't throw any exception and the loss was going down, and I assumed everything was going fine. 👎  Luckily, this code example is not shown in the book (which probably explains why I wasn't more careful).

Here are the errors:

1. The initial state tensor should have shape `[batch_size, self.cell.state_size]` instead of just `[self.cell.state_size]`
2. The initial outputs tensor should have shape `[batch_size, self.cell.output_size]` instead of `[n_steps, self.cell.output_size]`
3. The `sequences` array must be defined before the loop even when `self.return_sequences` is `False` (that's somewhat of a weird constraint imposed by `tf.function()`).
4. As you pointed out, `sequences.stack()` should be transposed.

Here's the corrected `MyRNN` class:

```python
class MyRNN(keras.layers.Layer):
    def __init__(self, cell, return_sequences=False, **kwargs):
        super().__init__(**kwargs)
        self.cell = cell
        self.return_sequences = return_sequences
        self.get_initial_state = getattr(
            self.cell, "get_initial_state", self.fallback_initial_state)
    def fallback_initial_state(self, inputs):
        batch_size = tf.shape(inputs)[0]
        return [tf.zeros([batch_size, self.cell.state_size], dtype=inputs.dtype)]
    @tf.function
    def call(self, inputs):
        states = self.get_initial_state(inputs)
        shape = tf.shape(inputs)
        batch_size = shape[0]
        n_steps = shape[1]
        sequences = tf.TensorArray(
            inputs.dtype, size=(n_steps if self.return_sequences else 0))
        outputs = tf.zeros(shape=[batch_size, self.cell.output_size], dtype=inputs.dtype)
        for step in tf.range(n_steps):
            outputs, states = self.cell(inputs[:, step], states)
            if self.return_sequences:
                sequences = sequences.write(step, outputs)
        if self.return_sequences:
            return tf.transpose(sequences.stack(), [1, 0, 2])
        else:
            return outputs
```

It now produces exactly the same output as a `keras.layers.RNN`, both when `return_sequences` is `True` and when it's `False`:

```python
for seq in [True, False]:
    cell = LNSimpleRNNCell(20)    
    my_rnn = MyRNN(cell, return_sequences=seq, input_shape=[None, 1])
    regular_rnn = keras.layers.RNN(cell, return_sequences=seq, input_shape=[None, 1])
    assert np.allclose(my_rnn(X_train), regular_rnn(X_train))
```

Regarding performance, I can't find a more efficient way to do `tf.transpose(sequences.stack(), [1, 0, 2])`. I'm seeing roughly the same performance with `keras.layers.RNN` and `MyRNN`.

Thanks again for your feedback. I'll push the fix now.
